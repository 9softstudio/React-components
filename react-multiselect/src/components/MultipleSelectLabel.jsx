import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';

const NO_SELECTED_LABEL = "Select options";
const MAX_SELECTED_ITEM_FOR_DISPLAY = 3;

export default class MultipleSelectLabel extends Component {
    constructor(props) {
        super(props);

        this.state = {
            noneSelectedLabel: props.noneSelectedLabel
        }
    }

    static propTypes = {
        selectedItems: PropTypes.arrayOf(PropTypes.object).isRequired,
        noneSelectedLabel: PropTypes.string,
        maxDisplayItemCount: PropTypes.number
    }

    static defaultProps = {
        noneSelectedLabel: NO_SELECTED_LABEL,
        maxDisplayItemCount: MAX_SELECTED_ITEM_FOR_DISPLAY
    }

    get selectedItemsString() {
        const { selectedItems, noneSelectedLabel, maxDisplayItemCount } = this.props;
        const selectedValues = selectedItems.map(item => item.value);

        let displayText = noneSelectedLabel;
        const selectedItemCount = selectedValues.length;

        if (selectedItemCount >= maxDisplayItemCount) {
            displayText = `${selectedItemCount} selected`;
        }
        else if (selectedItemCount >= 1) {
            displayText = selectedValues.join(", ");
        }

        return displayText;
    }

    render() {
        return (
            <button type="button" className="multiple-select-default multiple-select-label"
                onClick={this.props.onToggle}>
                <span className="text-display" title={this.selectedItemsString}>{this.selectedItemsString}</span>
                <b className="caret"></b>
            </button>
        )
    }
}