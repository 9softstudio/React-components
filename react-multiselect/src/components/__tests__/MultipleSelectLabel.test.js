import React from 'react'
import renderer from 'react-test-renderer'

import MultipleSelectLabel from '../MultipleSelectLabel'

const NO_SELECTED_LABEL = "Select options";
const MAX_SELECTED_ITEM_FOR_DISPLAY = 3;

const defaultSelectedItem = {
    key: "keyDefault",
    value: "valueDefault"
};
const defaultProps = {
    selectedItems: [],
    noneSelectedLabel: NO_SELECTED_LABEL,
    maxDisplayItemCount: MAX_SELECTED_ITEM_FOR_DISPLAY
}

describe('MultipleSelectLabel ', () => {
    afterAll(() => {
        jest.resetAllMocks();
    })

    describe('selectedItemsString', () => {
        it('No item is selected - return "Select options" text', () => {
            const expectedResult = "Select options";
            const label = new MultipleSelectLabel(defaultProps);

            const result = label.selectedItemsString;

            expect(result).toBe(expectedResult);
        })

        it('selectedItemCount >= maxDisplayItemCount - return selected item count text', () => {
            const expectedResult = "3 selected";
            const selectedItems = [defaultSelectedItem, defaultSelectedItem, defaultSelectedItem]

            const props = Object.assign({}, defaultProps, { selectedItems });
            const label = new MultipleSelectLabel(props);

            const result = label.selectedItemsString;

            expect(result).toBe(expectedResult);
        })

        it('has selected item and item count is less than maxDisplayItemCount - return value of selected items with comma separator', () => {
            const expectedResult = "valueDefault, value1";
            const selectedItems = [defaultSelectedItem, { key: "key1", value: "value1" }];

            const props = Object.assign({}, defaultProps, { selectedItems });
            const label = new MultipleSelectLabel(props);

            const result = label.selectedItemsString;

            expect(result).toBe(expectedResult);
        })
    })

    describe('render', () => {
        it('rendered content should match to snap shot', () => {
            const label = renderer.create(<MultipleSelectLabel {...defaultProps} />);

            const result = label.toJSON();

            expect(result).toMatchSnapshot();
        })

        it('click on label - props.onToggle should be called', () => {
            const toggleFunc = jest.fn();
            const props = Object.assign({}, defaultProps, { onToggle: toggleFunc });
            const label = renderer.create(<MultipleSelectLabel {...props} />);

            const renderedContent = label.toJSON();
            renderedContent.props.onClick();

            expect(toggleFunc).toHaveBeenCalled();
        })
    })
})