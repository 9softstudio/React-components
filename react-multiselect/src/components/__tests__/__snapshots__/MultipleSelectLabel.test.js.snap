// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultipleSelectLabel  render rendered content should match to snap shot 1`] = `
<button
  className="multiple-select-default multiple-select-label"
  onClick={undefined}
  type="button"
>
  <span
    className="text-display"
    title="Select options"
  >
    Select options
  </span>
  <b
    className="caret"
  />
</button>
`;
